lib/Dialect/FIRRTL/Transforms/CheckCombCycles.cpp

```diff
@@ -0,0 +1,470 @@
+      else if (fieldIndex == dataIndex)
+        dstNode = combGraph.getOrAddNode(use.get(), fieldIndex + 1);
+    }
+
+    // Now, we can create edges.
+    if (dstNode) {
+      for (auto srcNode : srcNodes)
+        srcNode->addChild(dstNode);
+    }
+  }
+  return true;
+}
+
+/// Expression handlers.
+bool Detector::visitExpr(SubfieldOp op) {
+  auto definingOp = op.input().getDefiningOp();
+  if (!definingOp || !isa<MemOp>(definingOp)) {
+    op->emitOpError("input must be a port of a MemOp, please run "
+                    "-firrtl-lower-types first");
+    return false;
+  }
+
+  auto portType = op.input().getType().cast<BundleType>();
+  auto fieldIndex = op.fieldIndex();
+  // If the memory port is not on a combinational path, it was not be inserted
+  // into the graph as a node in the MemOp handler.
+  if (auto memPortNode = combGraph.getNode(op.input(), fieldIndex + 1)) {
+    auto node = combGraph.getOrAddNode(op.result());
+    // We use `isFlip` to determine the direction of the edge.
+    if (portType.getElement(fieldIndex).isFlip)
+      memPortNode->addChild(node);
+    else
+      node->addChild(memPortNode);
+  }
+  return true;
+}
+
+/// Handle other operations that don't have concrete visitors.
+bool Detector::visitUnhandledOp(Operation *op) {
+  // Create an edge between each exist input node and each output node.
+  SmallVector<CombNode, 4> dstNodes;
+  for (auto result : op->getResults())
+    dstNodes.push_back(combGraph.getOrAddNode(result));
+
+  for (auto operand : op->getOperands()) {
+    if (auto srcNode = combGraph.getNode(operand)) {
+      for (auto dstNode : dstNodes)
+        srcNode->addChild(dstNode);
+    }
+  }
+  return true;
+}
+
+/// Build combinational graph of the module with the visitor dispatcher.
+void Detector::buildCombGraph() {
+  // Each module port should be a node in the graph.
+  for (auto arg : module.getArguments())
+    combGraph.getOrAddNode(arg);
+
+  // Traverse all operations in the module and create combinational nodes and
+  // edges.
+  module.walk([&](Operation *op) {
+    if (!dispatchVisitor(op))
+      return WalkResult::interrupt();
+    return WalkResult::advance();
+  });
+}
+
+/// Detect combinational cycles in the module. Also detect combinational paths
+/// between IOs of the module and record these paths in the `combPathsMap`.
+void Detector::detect() {
+  buildCombGraph();
+
+  // Add a dummy node in the begining to launch a Tarjan's SCC algorithm.
+  auto dummyNode = combGraph.getOrAddNode(nullptr);
+  for (auto connect : module.getOps<ConnectOp>()) {
+    if (auto node = combGraph.getNode(connect.dest()))
+      dummyNode->addChild(node);
+  }
+
+  // Lauch SCC iteration and report diagnostics.
+  using SCCIterator = llvm::scc_iterator<CombNode>;
+  for (auto SCC = SCCIterator::begin(dummyNode); !SCC.isAtEnd(); ++SCC) {
+    if (SCC.hasCycle()) {
+      auto errorDiag = mlir::emitError(
+          module.getLoc(), "detected combinational SCC in the module");
```

Comment:
Nit: plainer language would make this more useful to a front end user.  E.g., a chisel user who gets this error may have no idea what an SCC is. Consider:

```suggestion
          module.getLoc(), "detected combinational cycle in a FIRRTL module");
```

Note: the usual synthesis tool error is about "combinational loops" which is unfortunate.  It is also unfortunate that "cycle" has a collision with "clock cycle".  However, "cycle" is the correct mathematical term.